### ISSUE TITLE
[FEATURE] Add NULL DynamoDB Attribute Type

### ISSUE THREAD

**Issue by fishcharlie** [dynamoose/dynamoose#951]
### Summary:

Dynamoose should support the DynamoDB NULL Attribute Type.

### Other information (if applicable):

When we implement this we need to add the following test to `Document.isDynamoObject`:

```js
{
    "input": {"id": {"S": "foo" }, "data": {"NULL": true}},
    "output": true
}
```

Related Issues:

- #878 

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this feature has not already been suggested before
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

**Comment by uzochukwuonuegbu** [comment]
Hello @fishcharlie !

I have a similar issue, a null type attribute in dynamoDB doesn't allow me to fetch the document. Throws a typeMismatch error: TypeMismatch: Expected libraries to be of type array, instead found type object.

I believe supporting Null DynamoDB attribute type would solve this...right?

**Comment by fishcharlie** [comment]
@uzochukwuonuegbu Sounds like it, yes.

What is your use case here? One thing I'm trying to figure out is, how this should work. Because in my mind, there isn't **that** much of a difference between having the property not exist vs having the property value = null. So in your logic, how are you thinking about a property that doesn't exist vs it equaling null? What is the difference in your mind?

To me, they are almost the same.

To me there are a few different options here. For one, we could have a `null` Dynamoose type. And you can set it to allow a property to be `null`. If that type isn't defined as a valid type for the given attribute tho, it'll still fail with the same error you are getting now.

OR, we could have an attribute settings property that you could set to true to allow null values. Or have no setting and null properties just work with no code changes required (except if `required` is set on the property).

Lots of options here, all with pros and cons. So would love to hear more about your use case and what you are looking to do here, and why you need/want this functionality.

**Comment by fishcharlie** [comment]
[Comment from @williscool in regards to use case](https://github.com/dynamoose/dynamoose/issues/878#issuecomment-654645015):

> @fishcharlie Our use case is communicating with a third party that stores its own state.
Sending a property will update that property on their end, no property at all means 'leave as is', and a value of null will mean 'delete that property'.

> Another similar use case would be that in JSON you can't represent { foo: undefined }, but you can represent {} and { foo: null }.

> Thanks for looking into this and discovering the root cause!

**Comment by fahimmahmoodmir** [comment]
@fishcharlie I also had a similar problem like @williscool  that is for having some properties coming as null caused TypeMismatch error on hashKey. Our use case is that we want to have a property having `null` or  `""`  to differentiate that certain feature was never initialized and having `''` mean that certain feature was initialized but never used.

**Comment by fishcharlie** [comment]
@fahimmahmoodmir Just out of curiosity then, how are you treating a property not existing at all? Like if a property doesn't exist at all in the document, how is that treated?

**Comment by fahimmahmoodmir** [comment]
@fishcharlie currently we didn't have a case where the property is not existing

**Comment by felix-sg** [comment]
@fishcharlie we are having this issue as well.
In our documents there are nested objects that come from a different service.
We do not control the content nor want to apply any data validation.

Sometimes those services send us variables set to `null`.
The thing is just that we want to be able to save the object as it comes.

**Comment by eduardocalazansjr** [comment]
> @fishcharlie we are having this issue as well.
> In our documents there are nested objects that come from a different service.
> We do not control the content nor want to apply any data validation.
> 
> Sometimes those services send us variables set to `null`.
> The thing is just that we want to be able to save the object as it comes.

Same issue here.
I can save the WS response as a object on dynamodb... but then i can't retrieve this document anymore... it keeps throwing 'expect to be type string, instead found type object'

**Comment by ealiaga** [comment]
I had the same problem. I was able to safely save the document but couldn't get it. I got the error "**TypeMismatch: Expected key to be of type boolean, instead found type object.**"
Unfortunately it happened in **production** and I can't wait for an improvement in Dynamoose. I solved it with the [AWS SDK DynamoDB client.](https://github.com/dabit3/dynamodb-documentclient-cheat-sheet)

I think Dynamoose is a great library that delivers a better developer experience than the AWS SDK, but this issue forced me to migrate to the AWS SDK.
